apps/els_lsp/src/els_diagnostics_provider.erl

```diff
@@ -2,90 +2,44 @@
+-export([ is_enabled/0
         , options/0
+        , handle_request/2
         ]).
 
 -export([ notify/2
         , publish/2
         ]).
 
-
 %%==============================================================================
 %% Includes
 %%==============================================================================
 -include("els_lsp.hrl").
 -include_lib("kernel/include/logger.hrl").
 
--type state() :: #{in_progress => [progress_entry()]}.
--type progress_entry() :: #{ uri := uri()
-                           , pending := [job()]
-                           , diagnostics := [els_diagnostics:diagnostic()]
-                           }.
--type job() :: pid().
-
--define(SERVER, ?MODULE).
-
 %%==============================================================================
 %% els_provider functions
 %%==============================================================================
-
 -spec is_enabled() -> boolean().
 is_enabled() -> true.
 
 -spec options() -> map().
 options() ->
   #{}.
 
--spec init() -> state().
-init() ->
-  #{ in_progress => [] }.
-
-%% LSP 3.15 introduce versioning for diagnostics. Until all clients
-%% support it, we need to keep track of old diagnostics and re-publish
-%% them every time we get a new chunk.
--spec handle_info(any(), state()) -> state().
-handle_info({diagnostics, Diagnostics, Job}, State) ->
-  ?LOG_DEBUG("Received diagnostics [job=~p]", [Job]),
-  #{ in_progress := InProgress } = State,
-  { #{ pending := Jobs
-     , diagnostics := OldDiagnostics
-     , uri := Uri
-     }
-  , Rest
-  } = find_entry(Job, InProgress),
-  NewDiagnostics = Diagnostics ++ OldDiagnostics,
-  ?MODULE:publish(Uri, NewDiagnostics),
-  case lists:delete(Job, Jobs) of
-    [] ->
-      State#{in_progress => Rest};
-    Remaining ->
-      State#{in_progress => [#{ pending => Remaining
-                              , diagnostics => NewDiagnostics
-                              , uri => Uri
-                              }|Rest]}
-  end;
-handle_info(_Request, State) ->
-  State.
-
--spec handle_request(any(), state()) -> {any(), state()}.
-handle_request({run_diagnostics, Params}, State) ->
-  #{in_progress := InProgress} = State,
+-spec handle_request(any(), any()) -> {diagnostics, uri(), [pid()]}.
+handle_request({run_diagnostics, Params}, _State) ->
   #{<<"textDocument">> := #{<<"uri">> := Uri}} = Params,
   ?LOG_DEBUG("Starting diagnostics jobs [uri=~p]", [Uri]),
   Jobs = els_diagnostics:run_diagnostics(Uri),
-  Entry = #{uri => Uri, pending => Jobs, diagnostics => []},
-  {noresponse, State#{in_progress => [Entry|InProgress]}}.
+  {diagnostics, Uri, Jobs}.
 
 %%==============================================================================
 %% API
 %%==============================================================================
 -spec notify([els_diagnostics:diagnostic()], pid()) -> ok.
 notify(Diagnostics, Job) ->
-  ?SERVER ! {diagnostics, Diagnostics, Job},
+  els_provider ! {diagnostics, Diagnostics, Job},
```

Comment:
I wonder if those notifications should be somehow incorporated into the background job framework - e.g. that it sends as a reply to the process that started the job the value that was returned from the function (kind of similar to Elixir's `Task`)

Reply:
I'm not familiar with Elixir's `Task` but it sounds like a good idea. Created #1272